gcp/modules/locust/client.rb

```diff
@@ -2,79 +2,90 @@
+    unless distributions.empty?
+      distributions = distributions.reverse.first
+      ["100th_percentile", "99th_percentile", "98th_percentile", "95th_percentile", "90th_percentile",
+       "80th_percentile", "75th_percentile", "66th_percentile", "50th_percentile"].each do |metric|
+        metrics[metric] = distributions.pop
+      end
     end
+
+    return metrics
   end
 
-  metric_service_client = Google::Cloud::Monitoring::Metric.new(version: :v3)
-  formatted_name = Google::Cloud::Monitoring::V3::MetricServiceClient.project_path(@project_id)
-
-  time_series = []
-  time = Time.now.to_i
-
-  metrics.each do |metric, value|
-
-    time_series << {
-      "metric" => {
-        "type" => "custom.googleapis.com/locust.io/#{app_name}/#{metric}",
-      },
-      "resource" => {
-        "type" => "global",
-        "labels" => {
-          "project_id" => "#{@project_id}"
-        }
-      },
-      "metric_kind" => "GAUGE",
-      "value_type" => "DOUBLE",
-      "points" => [
-        {
-          "interval" => {
-            "end_time" => {
-              "seconds" => time,
-              "nanos" => 0
+  def self.process_locust_result(locust_stats_file, locust_distribution_file, app_name, user_count)
+    all_stats = JSON.parse(File.read(locust_stats_file))
+    all_distributions = CSV.parse(File.read(locust_distribution_file))
+    collected_metrics = collect_metrics(all_stats, all_distributions, user_count)
+
+    metric_service_client = Google::Cloud::Monitoring::Metric.new(version: :v3)
+    formatted_name = Google::Cloud::Monitoring::V3::MetricServiceClient.project_path(@project_id)
+
+    time_series = []
+    time = Time.now.to_i
+
+    collected_metrics.each do |metric, value|
+
+      time_series << {
+        "metric" => {
+          "type" => "custom.googleapis.com/locust.io/#{app_name}/#{metric}",
+        },
+        "resource" => {
+          "type" => "global",
+          "labels" => {
+            "project_id" => "#{@project_id}"
+          }
+        },
+        "metric_kind" => "GAUGE",
+        "value_type" => "DOUBLE",
+        "points" => [
+          {
+            "interval" => {
+              "end_time" => {
+                "seconds" => time,
+                "nanos" => 0
+              }
+            },
+            "value" => {
+              "double_value" => value.to_f
             }
-          },
-          "value" => {
-            "double_value" => value.to_f
           }
-        }
-      ]
-    }
-  end
+        ]
+      }
+    end
 
-  begin
-    metric_service_client.create_time_series(formatted_name, time_series)
-  rescue Google::Gax::RetryError
-    puts "[ERROR]: Error while submitting metrics to Stackdriver (Google::Gax::RetryError)."
-    exit 120
+    begin
+      metric_service_client.create_time_series(formatted_name, time_series)
+    rescue Google::Gax::RetryError
+      puts "[ERROR]: Error while submitting metrics to Stackdriver (Google::Gax::RetryError)."
```

Comment:
I know this did not change, but what about printing some more details about the error (in the light of same error being thrown in the Stackdriver mgmt. code when Stackdriver workspace is not present)? (seems to provide some extra info - https://www.rubydoc.info/gems/google-gax/Google/Gax/RetryError)

Reply:
I added a print and tested it manually with a dumb RetryError I made and it works, but I'm not sure how it will look for a "real" RetryError (because I'm not sure of an easy way to trigger a "real" failure here).